### ISSUE TITLE
Support for containers

### ISSUE THREAD

**Issue by martinthomson** [bwinton/VerticalTabs#405]
The current UX for containers uses a coloured band at the top of the tab to indicate which container is active for that tab.

**Comment by Work-Parker** [comment]
Still having this issue.
Firefox Nightly on Windows 10 64 bit

******START OF ABOUT:SUPPORT INFORMATION******
Application Basics
------------------

Name: Firefox
Version: 51.0a1
Build ID: 20160804030441
Update Channel: nightly
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:51.0) Gecko/20100101 Firefox/51.0
OS: Windows_NT 10.0
Multiprocess Windows: 1/1 (Enabled by default)
Safe Mode: false

Crash Reports for the Last 3 Days
---------------------------------

Report ID: bp-a0abf4d3-d12e-4678-9ad1-199d82160805
Submitted: 14 minutes ago

Report ID: bp-c157eb6a-27c6-4fef-b06f-9cc852160803
Submitted: 2 days ago

Report ID: bp-f82be0f5-17fd-4a48-9e22-3cde12160803
Submitted: 2 days ago

All Crash Reports

Extensions
----------

Name: Activity Stream
Version: 1.1.1
Enabled: true
ID: @activity-streams

Name: FlyWeb
Version: 1.0.0
Enabled: true
ID: flyweb@mozilla.org

Name: Multi-process staged rollout
Version: 1.0
Enabled: true
ID: e10srollout@mozilla.org

Name: No More 404s
Version: 1.5.3
Enabled: true
ID: wayback_machine@mozilla.org

Name: Pocket
Version: 1.0.4
Enabled: true
ID: firefox@getpocket.com

Name: Tab Center
Version: 1.23
Enabled: true
ID: tabcentertest1@mozilla.com

Name: Test Pilot
Version: 0.8.3-tag-2016-07-29
Enabled: true
ID: @testpilot-addon

Name: Web Compat
Version: 1.0
Enabled: true
ID: webcompat@mozilla.org

Name: uBlock Origin
Version: 1.8.0
Enabled: false
ID: uBlock0@raymondhill.net

Name: Universal Search
Version: 1.0.8
Enabled: false
ID: universal-search@mozilla.com

Graphics
--------

Features
Compositing: Direct3D 11
Asynchronous Pan/Zoom: wheel input enabled; touch input enabled
WebGL Renderer: Google Inc. -- ANGLE (NVIDIA Quadro NVS 160M Direct3D11 vs_4_0 ps_4_0)
WebGL2 Renderer: WebGL creation failed: * Refused to create native OpenGL context because of blacklist entry: WEBGL_NATIVE_GL_OLD_NVIDIA * Exhausted GL driver options.
Hardware H264 Decoding: No; Hardware video decoding disabled or blacklisted
Audio Backend: wasapi
Direct2D: true
DirectWrite: true (10.0.10586.494)
GPU #1
Active: Yes
Description: NVIDIA Quadro NVS 160M
Vendor ID: 0x10de
Device ID: 0x06eb
Driver Version: 9.18.13.4192
Driver Date: 10-13-2015
Drivers: nvd3dumx,nvwgf2umx,nvwgf2umx nvd3dum,nvwgf2um,nvwgf2um
Subsys ID: 024f1028
RAM: 256

Diagnostics
AzureCanvasAccelerated: 0
AzureCanvasBackend: direct2d 1.1
AzureContentBackend: direct2d 1.1
AzureFallbackCanvasBackend: cairo





Important Modified Preferences
------------------------------

accessibility.typeaheadfind.flashBar: 0
browser.cache.disk.capacity: 358400
browser.cache.disk.filesystem_reported: 1
browser.cache.disk.smart_size.first_run: false
browser.cache.disk.smart_size.use_old_max: false
browser.cache.frecency_experiment: 4
browser.download.importedFromSqlite: true
browser.download.useDownloadDir: false
browser.places.smartBookmarksVersion: 8
browser.sessionstore.upgradeBackup.latestBuildID: 20160804030441
browser.startup.homepage: resource://activity-streams/data/content/activity-streams.html#/
browser.startup.homepage_override.buildID: 20160804030441
browser.startup.homepage_override.mstone: 51.0a1
browser.tabs.animate: false
browser.tabs.crashReporting.email: parkerhasmail@protonmail.com
browser.tabs.crashReporting.emailMe: true
browser.tabs.crashReporting.includeURL: true
browser.tabs.drawInTitlebar: false
browser.urlbar.suggest.searches: true
dom.apps.lastUpdate.buildID: 20160725030248
dom.apps.lastUpdate.mstone: 50.0a1
dom.apps.reset-permissions: true
dom.mozApps.used: true
extensions.lastAppVersion: 51.0a1
font.internaluseonly.changed: false
gfx.crash-guard.d3d11layers.appVersion: 45.2.0
gfx.crash-guard.d3d11layers.deviceID: 0x06eb
gfx.crash-guard.d3d11layers.driverVersion: 9.18.13.4192
gfx.crash-guard.d3d11layers.feature-d2d: true
gfx.crash-guard.d3d11layers.feature-d3d11: true
gfx.crash-guard.status.d3d11layers: 2
gfx.crash-guard.status.d3d9video: 2
gfx.direct3d.last_used_feature_level_idx: 1
media.benchmark.vp9.fps: 61
media.benchmark.vp9.versioncheck: 1
media.gmp-eme-adobe.abi: x86_64-msvc-x64
media.gmp-eme-adobe.lastUpdate: 1470081446
media.gmp-eme-adobe.version: 17
media.gmp-gmpopenh264.abi: x86_64-msvc-x64
media.gmp-gmpopenh264.lastUpdate: 1470081447
media.gmp-gmpopenh264.version: 1.5.3
media.gmp-manager.buildID: 20160804030441
media.gmp-manager.lastCheck: 1470324903
media.gmp-widevinecdm.abi: x86_64-msvc-x64
media.gmp-widevinecdm.lastUpdate: 1469110883
media.gmp-widevinecdm.version: 1.4.8.866
media.gmp.storage.version.observed: 1
media.hardware-video-decoding.failed: true
network.cookie.cookieBehavior: 1
network.cookie.prefsMigrated: true
network.dns.disablePrefetch: true
network.http.pipelining.ssl: true
network.predictor.cleaned-up: true
places.database.lastMaintenance: 1469805250
places.favicons.optimizeToDimension: 64
places.history.expiration.transient_current_max_pages: 122334
plugin.disable_full_page_plugin_for_types: application/pdf
plugin.importedState: true
privacy.clearOnShutdown.cache: false
privacy.clearOnShutdown.downloads: false
privacy.clearOnShutdown.history: false
privacy.clearOnShutdown.offlineApps: true
privacy.cpd.cookies: false
privacy.sanitize.migrateClearSavedPwdsOnExit: true
privacy.sanitize.migrateFx3Prefs: true
privacy.sanitize.sanitizeOnShutdown: true
privacy.trackingprotection.enabled: true
privacy.trackingprotection.introCount: 20
security.sandbox.content.tempDirSuffix: {511db220-75e5-40d8-8658-4f9b73cf1298}
security.ssl.errorReporting.automatic: true
security.tls.version.max: 1
security.tls.version.min: 0
services.sync.declinedEngines:
storage.vacuum.last.index: 1
storage.vacuum.last.places.sqlite: 1469111303
ui.osk.debug.keyboardDisplayReason: IKPOS: Touch screen not found.

Important Locked Preferences
----------------------------

Places Database
---------------

JavaScript
----------

Incremental GC: true

Accessibility
-------------

Activated: false
Prevent Accessibility: 0

Library Versions
----------------

NSPR
Expected minimum version: 4.12
Version in use: 4.12

NSS
Expected minimum version: 3.26
Version in use: 3.26

NSSSMIME
Expected minimum version: 3.26
Version in use: 3.26

NSSSSL
Expected minimum version: 3.26
Version in use: 3.26

NSSUTIL
Expected minimum version: 3.26
Version in use: 3.26

Experimental Features
---------------------
******END OF ABOUT:SUPPORT INFORMATION******

**Comment by bwinton** [comment]
Yep.  No-one has worked on it, and it's not in the roadmap, so it's unlikely to be fixed anytime soon…  But you can follow this bug to see when it does get fixed.  🙂

**Comment by Work-Parker** [comment]
Any way I can signal boost it or whatever so they can put this on the roadmap?

**Comment by bwinton** [comment]
I'm the person responsible for the roadmap, so, uh, not really…  😉

(Containers is a really cool technology, but the current usage numbers are so low that it's just not as important as the other things I need to do.  If someone submitted a pull request that added container support, I wouldn't be opposed to merging it, but I don't see having the time to work on it myself in the near future.  Which probably isn't the answer you were hoping for, but I wanted to be honest with you.)

**Comment by Work-Parker** [comment]
Not really, but I don't mind. Just trying to help out where I can, and if there are more important issues that need addressing then we should defiantly help with those first. Can you point me where I should go so I can help with those?

**Comment by bwinton** [comment]
Possibly…  What kind of work are you looking to do?  (Did you want to try adding container support and submitting a pull request?)

**Comment by Work-Parker** [comment]
I'm new to github, so I don't really know what a pull request is, but if that would help, then sure.

**Comment by bwinton** [comment]
Cool, uh, email me (bwinton@mozilla.com), and I'll walk you through the process of contributing (and we can avoid sending Martin more email)…

**Comment by alex-mayorga** [comment]
¡Hola @bwinton!

This one bite me too =)

Where in the code shall I start looking at?

¡Gracias!
Alex

**Comment by bwinton** [comment]
Well, I think the first thing we would need was a design for how it should look.
The CSS for the tabs is mostly in https://github.com/bwinton/TabCenter/blob/master/skin/base.css#L79 if you wanted to start playing around with that…

**Comment by fvsch** [comment]
@bwinton I'd like to experiment with this. I was wondering if there might be a chance to group tabs by container, and inject a header element in the tab list. Do you think the current architecture could be extended that way, or would that go in a "hell no, requires a complete rewrite" category?

I was thinking about having:

```
Tab in general container
Tab in general container
[Container A - new tab - close all]
Tab in container A
Tab in container A
Tab in container A
[Container B - new tab - close all]
Tab in container B
Tab in container B
```

(Starting with grouping + headers without buttons)

**Comment by bwinton** [comment]
Soooo, I think you _might_ be able to extend the architecture that way, but the tabbrowser-tabs we're extending is remarkably finicky.  (There's a reason we don't have the pinned tabs in a horizontal row above the un-pinned tabs.  :cry:)  The other thing to note is that @ericawright is currently re-writing a bit of that code to make some of the reverse code easier (hopefully), so you may want to wait until [her PR](https://github.com/bwinton/TabCenter/pull/887) has landed (or base your code off of it, once she's pushed a more recent version).

But having said that, I think a number of people would be very happy if you took a stab at it, and Erica and I would be more than happy to help you out if you run into problems.  🙂

So, to get you started, here are a couple of things you'll need to consider:
1. We have a `reverse` mode.  Should that reverse just the tabs in each container, or the order of the containers as well?
2. What happens when you drag a tab from one container into another?  How do you prevent that?
3. Can pinned tabs be in a container?  How does that work?
4. When changing tabs with the keyboard, we scroll to the new tab.  Should we also show the container header if it's the first tab in a container?  What if it's the second?
5. When filtering with the searchbox at the top, should we include container names?  What if none of the tabs in the container match?  I think it might be neat to have a button that showed all the tabs in a container if you were filtering and clicked the container name.
6. I worry a little about trying to fit both a new tab button and a close all button in the container row…

Good luck!

**Comment by fvsch** [comment]
Good points re. grouping and not being able (by design) to change a tab from a container to another. I'll start smaller with just styling, to get parity with the horizontal tabs, but will try to think about that and experiment a bit.

On the technical side, I'd hoped the vertical tabs would have jumped on the React-or-similar bandwaggon has some devtools parts have, but sadly it's still good old gzool with the Flintstone's Flexbox. ^^ That's going to limit what's doable. Well, we'll see.

**Comment by bwinton** [comment]
Yeah, we would have liked to use something more modern, but not using the built-in tabstrip breaks everything, so that wasn't an option.  :cry:

**Comment by fvsch** [comment]
So, minimal step = color coding or some way to surface a container in each tab, like for horizontal tabs.

I’m getting this result, but I’m not very happy about it. Thoughts?
https://dl.dropboxusercontent.com/u/145744/tabcenter-containers-test.mov

**Comment by bwinton** [comment]
Ack, sorry for the delay…

I agree it's not great, particularly since you can't see what container a tab is in before switching to it.  But it's also better than nothing…  Let's see what @phlsa thinks!

**Comment by fvsch** [comment]
I tried earmarking tabs (very crudely). Might work better on the left.

<img width="275" alt="screen shot 2017-02-08 at 23 45 13" src="https://cloud.githubusercontent.com/assets/243601/22760966/1018b182-ee59-11e6-9d6c-8bfeda9ac70c.png">

I also tried showing the container’s icon, using code from the URL bar, but it seems the `--identity-icon-color` color is set on tabs, but not the `--identity-icon` (an image). I’ll have to look deeper.

**Comment by bwinton** [comment]
I think the left would be better…  What about making the background colour slightly tinted to the colour of the container?  Like at 10% opacity or something?

**Comment by benwaffle** [comment]
Yes, Why not just change the color of the blue line on the left?

**Comment by bwinton** [comment]
Because that only shows which container the current tab is in, not any of the others?  (I mean, I kind of like that idea, but it does lose information…)

**Comment by benwaffle** [comment]
add it to all of the tabs, and use the different background (or something else) to indicate current tab

**Comment by callahad** [comment]
Or an additional thin, colored line on all tabs to represent the container that embiggens when a tab is selected?

**Comment by Keith94** [comment]
This is how the container tabs currently look with the [Containers Experiment](https://testpilot.firefox.com/experiments/containers) installed.

![image](https://cloud.githubusercontent.com/assets/5490615/23524599/27c8dbd4-ff40-11e6-8de2-4bab136abd12.png)

**Comment by benwaffle** [comment]
It needs some improvement for the current tab though
![screenshot from 2017-03-02 15-06-52](https://cloud.githubusercontent.com/assets/1713819/23524731/e7c7012c-ff59-11e6-8ec3-96ba35f55667.png)

**Comment by benwaffle** [comment]
it needs some improvement for the current tab though
![screenshot from 2017-03-02 15-07-45](https://cloud.githubusercontent.com/assets/1713819/23524763/0d619546-ff5a-11e6-9513-a64bda18680c.png)

**Comment by bwinton** [comment]
@callahad: I think, now that #944 landed, it should be easier to change the colour of that bar, if the containers peeps wanted to fix it again.  :wink:

**Comment by fvsch** [comment]
It's actually https://github.com/bwinton/TabCenter/pull/963 which landed and has relevant changes. (https://github.com/bwinton/TabCenter/pull/944 had a lot of things so we closed it because of the risk of regressions).

With the current code (from 963), changing the highlight color would require something like:

```css
/* Use the Container color if available */
.tabbrowser-tab[usercontextid] {
  --tab-highlight-color: var(--identity-tab-color);
}

/* Make the indicator partly visible for Container tabs */
.tabbrowser-tab[usercontextid] .tab-selected-indicator:not([pinned]) {
  background-position: -3px 0;
}
```

It would need a few tests to make sure the animation is not broken in specific cases, which means checking there is no conflict with this: https://github.com/bwinton/TabCenter/commit/c14dce1f5ba6c6c68f7c6f2dbef2795c75b38a49

**Comment by grahamperrin** [comment]
For reference: 

- https://github.com/mozilla/testpilot-containers/issues/336 (2017-03-04) Embrace the use of containers for tab management
- [Sea Containers sidebar for managing tabs - Test Pilot / Containers - Mozilla Discourse](https://discourse.mozilla-community.org/t/sea-containers-sidebar-for-managing-tabs/15800?u=grahamperrin) (2017-05-11)
- https://github.com/mozilla/testpilot-containers/issues/504 (2017-05-11) Consider shipping sidebar tabs inside testpilot experiment
